src/osd/PG.cc

```diff
@@ -1877,6 +1877,7 @@ void PG::activate(ObjectStore::Transaction& t,
     pg_log.reset_recovery_pointers();
   } else {
     dout(10) << "activate - not complete, " << missing << dendl;
+    info.stats.stats.sum.num_objects_missing = missing.num_missing();
```

Comment:
Hmm, i think this should be num_objects_missing_on_primary.. and I think (?) that num_objects_missing isn't updated correctly anywhere?

Reply:
> I think (?) that num_objects_missing isn't updated correctly anywhere?

Exactly, that's why I try to make this member can trace each peer's number of missing objects correctly here. So in the next peering circle we could leverage this (and plus the log difference) to choose a more suitable or worth-recovering peers as acting.

Reply:
Hmm, looking a bit more closely, I think all we need is a call to publish_stats_to_osd() at the end of activate()?

Reply:
Nope. What I am intending to do is to let each peer to record its number of missing objects separately. publish_stats_to_osd (and later call to _update_calc_stats() ) will only update primary's in-memory peer-info's num_objects_missing.